File: Fluid Text Adventure/Fluid Text Adventure/Items/Sword.cs
using System;
using System.Collections.Generic;
using FluidHTN;
using FluidHTN.Compounds;
using Fluid_Text_Adventure.Screens;

namespace Fluid_Text_Adventure
{
    public class Sword : Item
    {
        public override string Description =>
            "On the ground lies a rusty, old sword.";


        public Sword(Func<AIContext, GoalState, GameScreen> SetGoal) : base(DefineDomain())
        {
            KeywordsToAction = new Dictionary<List<string>, Func<AIContext, GameScreen>>
            {
                {
                    new List<string> { "get", "sword" },
                    ctx => SetGoal?.Invoke(ctx, GoalState.GetSword)
                },
                {
                    new List<string> { "slash" },
                    ctx => SetGoal?.Invoke(ctx, GoalState.SlashAir)
                },
            };
        }

        private static Domain<AIContext> DefineDomain()
        {
            var getSwordDomain = new DomainBuilder<AIContext>("Get Sword Sub-domain")
                .Select("Get Sword")
                    .Condition("GOAL: Get Sword", (ctx => ctx.HasGoal(GoalState.GetSword)))
                    .Action("Get Sword")
                        .Condition("Has NOT Weapon", ctx => ctx.HasState(AIWorldState.HasWeapon, false))
                        .Do(Actions.GetSword)
                        .Effect("Has Weapon", EffectType.PlanAndExecute, (ctx, type) => ctx.SetState(AIWorldState.HasWeapon, true, type))
                        .Effect("Try Complete Goal", EffectType.PlanAndExecute, (ctx, type) => Actions.TryCompleteGoal(ctx, GoalState.GetSword, type))
                    .End()
                .End()
                .Build();

            var slashAirActionDomain = new DomainBuilder<AIContext>("Slash Air Action Sub-domain")
                .Action("Slash Air")
                    .Condition("GOAL: Slash Air", (ctx => ctx.HasGoal(GoalState.SlashAir)))
                    .Condition("Has Weapon", ctx => ctx.HasState(AIWorldState.HasWeapon))
                    .Do((ctx => Actions.Write(ctx, "You slash your sword through the air elegantly!")))
                    .Effect("Try Complete Goal", EffectType.PlanAndExecute, (ctx, type) => Actions.TryCompleteGoal(ctx, GoalState.SlashAir, type))
                .End()
                .Build();

            return new DomainBuilder<AIContext>("Sword sub-domain")
                .Splice(getSwordDomain)
                .Action("Already has sword")
                    .Condition("GOAL: Get Sword", (ctx => ctx.HasGoal(GoalState.GetSword)))
                    .Condition("Has Weapon", ctx => ctx.HasState(AIWorldState.HasWeapon))
                    .Do((ctx => Actions.Write(ctx, "But you're already wielding the sword!")))
                    .Effect("Complete Goal", EffectType.PlanAndExecute, (ctx, type) => ctx.SetGoal(GoalState.None, true, type))
                .End()
                .Select("Slash")
                    .Condition("GOAL: Slash Air", (ctx => ctx.HasGoal(GoalState.SlashAir)))
                    .Splice(slashAirActionDomain)
                    .Sequence("Pick up sword, then slash with it")
                        .Action("Temporary change goal")
                            .Effect("Get Sword Goal", EffectType.PlanOnly, (ctx, type) => Actions.ChangeGoal(ctx, type, GoalState.GetSword))
                        .End()
                        .Splice(getSwordDomain)
                        .Action("Temporary change goal")
                            .Effect("Slash Air Goal", EffectType.PlanOnly, (ctx, type) => Actions.ChangeGoal(ctx, type, GoalState.SlashAir))
                        .End()
                        .Splice(slashAirActionDomain)
                    .End()
                .End()
                .Build();
        }
    }
}

File: Fluid Text Adventure/FluidTextAdventureCPP/Items/Sword.cpp
#include "StdIncludes.h"
#include "CoreIncludes/BaseDomainBuilder.h"
#include "Actions.h"
#include "Sword.h"

SharedPtr<FluidHTN::Domain> Sword::DefineDomain()
{

    SharedPtr<FluidHTN::Domain> getSwordDomain;
    SharedPtr<FluidHTN::Domain> dropSwordDomain;
    SharedPtr<FluidHTN::Domain> slashAirActionDomain;
    // clang-format off
    {
    BaseDomainBuilder getSwordDomainBuilder("Get Sword Sub-domain");
        getSwordDomainBuilder.AddSelector("Get Sword");
			getSwordDomainBuilder.AddCondition("GOAL: Get Sword",
											   [](IContext& ctx) { return static_cast<AIContext&>(ctx).HasGoal(GoalState::GetSword); });
			getSwordDomainBuilder.AddAction("Get Sword");
                getSwordDomainBuilder.AddCondition("Has NOT Weapon",[](IContext& ctx) {
                     return static_cast<AIContext&>(ctx).HasStateAIWS(AIWorldState::HasWeapon,false);
				});
				getSwordDomainBuilder.AddOperator([](IContext& ctx) { return Actions::GetSword(static_cast<AIContext&>(ctx)); });
                getSwordDomainBuilder.AddEffect("Has Weapon", EffectType::PlanAndExecute, [](IContext& ctx, EffectType type) {
                     static_cast<AIContext&>(ctx).SetStateAIWS(AIWorldState::HasWeapon, true, type);
				});
                getSwordDomainBuilder.AddEffect("Try Complete Goal",
                        EffectType::PlanAndExecute,[](IContext& ctx, EffectType type) {
                         Actions::TryCompleteGoal(static_cast<AIContext&>(ctx), GoalState::GetSword,type);
				});
			getSwordDomainBuilder.End();
		getSwordDomainBuilder.End();

        getSwordDomain = getSwordDomainBuilder.Build();
    }
	{
            BaseDomainBuilder dropSwordDomainBuilder("Drop Sword Sub-domain");
			  dropSwordDomainBuilder.AddSelector("Drop Sword");
				  dropSwordDomainBuilder.AddCondition("GOAL: Drop Sword",  
											   [](IContext& ctx) { return static_cast<AIContext&>(ctx).HasGoal(GoalState::DropSword); });
				  dropSwordDomainBuilder.AddAction("Drop Sword");
					  dropSwordDomainBuilder.AddCondition("Has Weapon",[](IContext& ctx) {
						 return static_cast<AIContext&>(ctx).HasState(AIWorldState::HasWeapon);
					  });
					  dropSwordDomainBuilder.AddOperator([](IContext& ctx) { return Actions::DropSword(static_cast<AIContext&>(ctx)); });
					  dropSwordDomainBuilder.AddEffect("Has NOT Weapon",
                                          EffectType::PlanAndExecute,[](IContext& ctx, EffectType type) {
                                           static_cast<AIContext&>(ctx).SetStateAIWS(AIWorldState::HasWeapon, false, type);
					   });
						dropSwordDomainBuilder.AddEffect("Try Complete Goal",
								EffectType::PlanAndExecute,[](IContext& ctx, EffectType type) {
								 Actions::TryCompleteGoal(static_cast<AIContext&>(ctx), GoalState::DropSword,type);
						});
				  dropSwordDomainBuilder.End();
			  dropSwordDomainBuilder.End();
			dropSwordDomain = dropSwordDomainBuilder.Build();
	}
    {
	    BaseDomainBuilder slashAirActionDomainBuilder("Slash Air Action Sub-domain");
        slashAirActionDomainBuilder.AddAction("Slash Air");
		   slashAirActionDomainBuilder.AddCondition("GOAL: Slash Air", 
						   [](IContext& ctx) { return static_cast<AIContext&>(ctx).HasGoal(GoalState::SlashAir); });
			slashAirActionDomainBuilder.AddCondition("Has Weapon",[](IContext& ctx) {
				 return static_cast<AIContext&>(ctx).HasState(AIWorldState::HasWeapon);
			});
			slashAirActionDomainBuilder.AddOperator([](IContext& ctx) { return Actions::Write(static_cast<AIContext&>(ctx), "You slash your sword through the air elegantly!");});
			slashAirActionDomainBuilder.AddEffect("Try Complete Goal",
					EffectType::PlanAndExecute,[](IContext& ctx, EffectType type) {
					 Actions::TryCompleteGoal(static_cast<AIContext&>(ctx), GoalState::SlashAir,type);
			 });
		slashAirActionDomainBuilder.End();
        slashAirActionDomain = slashAirActionDomainBuilder.Build();

	}
    {
        BaseDomainBuilder swordDomainBuilder("Sword sub-domain");
            swordDomainBuilder.Splice(*getSwordDomain);
            swordDomainBuilder.Splice(*dropSwordDomain);
            swordDomainBuilder.AddAction("Already has sword");
				swordDomainBuilder.AddCondition("GOAL: Get Sword", 
						   [](IContext& ctx) { return static_cast<AIContext&>(ctx).HasGoal(GoalState::GetSword); });
				swordDomainBuilder.AddCondition("Has Weapon", [](IContext& ctx) {
					 return static_cast<AIContext&>(ctx).HasState(AIWorldState::HasWeapon);
				});
				swordDomainBuilder.AddOperator([](IContext& ctx) { return Actions::Write(static_cast<AIContext&>(ctx), "But you're already wielding the sword!");});
				swordDomainBuilder.AddEffect("Complete Goal", EffectType::PlanAndExecute, [](IContext& ctx, EffectType type) {
					 static_cast<AIContext&>(ctx).SetGoal(GoalState::None, true, type);
				 });
            swordDomainBuilder.End();
            swordDomainBuilder.AddAction("Not holding sword to drop");
				swordDomainBuilder.AddCondition("GOAL: Drop Sword",
							   [](IContext& ctx) { return static_cast<AIContext&>(ctx).HasGoal(GoalState::DropSword); });
				swordDomainBuilder.AddCondition("Has NOT Weapon", [](IContext& ctx) {
						 return static_cast<AIContext&>(ctx).HasStateAIWS(AIWorldState::HasWeapon,false);
				});
				swordDomainBuilder.AddOperator([](IContext& ctx) { return Actions::Write(static_cast<AIContext&>(ctx), "But you're not holding a sword!");});
				swordDomainBuilder.AddEffect("Complete Goal", EffectType::PlanAndExecute, [](IContext& ctx, EffectType type) {
					 static_cast<AIContext&>(ctx).SetGoal(GoalState::None, true, type);
				 });
            swordDomainBuilder.End();
            swordDomainBuilder.AddSelector("Slash");
				swordDomainBuilder.AddCondition("GOAL: Slash Air", 
							   [](IContext& ctx) { return static_cast<AIContext&>(ctx).HasGoal(GoalState::SlashAir); });
				swordDomainBuilder.Splice(*slashAirActionDomain);
				swordDomainBuilder.AddSequence("Pick up sword, then slash with it");
				swordDomainBuilder.AddAction("Temporary change goal");
					swordDomainBuilder.AddEffect("Get Sword Goal", EffectType::PlanOnly, [](IContext& ctx, EffectType type) {
						 Actions::ChangeGoal(static_cast<AIContext&>(ctx), type, GoalState::GetSword);
					});
				swordDomainBuilder.End();
				swordDomainBuilder.Splice(*getSwordDomain);
				swordDomainBuilder.AddAction("Temporary change goal");
						swordDomainBuilder.AddEffect("Slash Air Goal", EffectType::PlanOnly,  [](IContext& ctx, EffectType type) {
						 Actions::ChangeGoal(static_cast<AIContext&>(ctx), type, GoalState::GetSword);
						});
				swordDomainBuilder.End();
				swordDomainBuilder.Splice(*slashAirActionDomain);
            swordDomainBuilder.End();
		swordDomainBuilder.End();

		return swordDomainBuilder.Build();
	}
}
